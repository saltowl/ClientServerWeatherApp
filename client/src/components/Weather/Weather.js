import React from "react";
import Loader from "../Loader/Loader";
import "./Weather.scss";

class Weather extends React.Component {
  constructor(props) {
    super(props);

    if (typeof this.props.data !== 'undefined') {
      this.state = {
        pending:
          typeof this.props.data.pending !== "undefined"
            ? this.props.data.pending
            : true,
        isError: typeof this.props.data.error !== "undefined",
        isCurrentCity: this.props.data.id === 0
      };
    } else {
      this.state = {
        isError: false,
        isCurrentCity: true,
        error: 'Cannot get current city'
      };
    }
  }

  componentDidMount() {
    this.props.fetchWeather();
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (
      typeof prevState.error === 'undefined' &&
      typeof nextProps.data.pending !== "undefined" &&
      nextProps.data.pending !== prevState.pending
    ) {
      return {
        pending: nextProps.data.pending,
        isError: typeof nextProps.data.error !== "undefined"
      };
    } else return null;
  }

  render() {
    const cityInWeatherList = !this.state.isCurrentCity;
    
    const styleWeather = cityInWeatherList
      ? { flexDirection: "column" }
      : { flexDirection: "row", width: "100%", fontSize: "30px" };
    const styleMainInfo = cityInWeatherList
      ? { flexDirection: "row" }
      : { flexDirection: "column", width: "50%", paddingRight: "20px" };
    const dataRow = cityInWeatherList
      ? { width: "100%" }
      : { width: "50%", paddingLeft: "20px", boxSizing: "border-box" };

    if (this.state.isError) {
      const error = this.props.data.error;
      let notFound = false;
      if (typeof error.response !== "undefined") {
        notFound = error.response.status === 404;
      }
      return (
        <div className="Weather error">
          {notFound ? (
            <div>{this.props.data.name} isn't found</div>
          ) : (
            <div>Poor connection. Please, try again later</div>
          )}
          {cityInWeatherList && (
            <button onClick={this.props.deleteCity} className="circle">
              x
            </button>
          )}
        </div>
      );
    } else if (!this.state.pending && typeof this.state.error === 'undefined') {
      const data = this.props.data.weather;

      const temp = `${data.main.temp < 0 ? "" : "+"}${data.main.temp.toFixed(
        1
      )}°C`;
      const icon = !this.state.isCurrentCity
        ? `https://openweathermap.org/img/wn/${data.weather[0].icon}.png`
        : `https://openweathermap.org/themes/openweathermap/assets/vendor/owm/img/widgets/${data.weather[0].icon}.png`;
      const description =
        data.weather[0].description[0].toUpperCase() +
        data.weather[0].description.substr(1);
      const rot = `rotate(${data.wind.deg}deg)`;
      const windDirStyle = {
        WebkitTransform: rot,
        msTransform: rot,
        OTransform: rot,
        transform: rot
      };
      const windSpeed = `${data.wind.speed} m/s`;
      const pressure = `${(data.main.pressure * 0.75006375541921).toFixed(
        0
      )} mmHg`;
      const humidity = `${data.main.humidity}%`;
      const coords = [data.coord.lon, data.coord.lat];

      return (
        <div className="Weather" style={styleWeather}>
          <div className="mainInfo" style={styleMainInfo}>
            <div className={this.state.isCurrentCity ? "currentName" : "name"}>
              {data.name}
            </div>
            <div
              className={`${
                this.state.isCurrentCity ? "current" : ""
              } cityMain`}
            >
              <div className={this.state.isCurrentCity ? "temp" : ""}>
                {temp}
              </div>
              <img src={icon} title={description} />
            </div>
            <div className={this.state.isCurrentCity ? "current" : ""}>
              {cityInWeatherList && (
                <button onClick={this.props.deleteCity} className="circle">
                  x
                </button>
              )}
            </div>
          </div>
          <div style={dataRow}>
            <div className="row">
              <div className="rowKey">Wind:</div>
              <div className="rowVal">
                <div id="wind">
                  <div id="N">N</div>
                  <div id="W">W</div>
                  <div id="windDirection">
                    <div style={windDirStyle}>↑</div>
                  </div>
                  <div id="E">E</div>
                  <div id="S">S</div>
                </div>
                {windSpeed}
              </div>
            </div>
            <div className="row">
              <div className="rowKey">Clouds:</div>
              <div className="rowVal">{data.clouds.all}%</div>
            </div>
            <div className="row">
              <div className="rowKey">Pressure:</div>
              <div className="rowVal">{pressure}</div>
            </div>
            <div className="row">
              <div className="rowKey">Humidity:</div>
              <div className="rowVal">{humidity}</div>
            </div>
            <div className="row">
              <div className="rowKey">Coords:</div>
              <div className="rowVal">
                [{coords[0]}, {coords[1]}]
              </div>
            </div>
          </div>
        </div>
      );
    } else if (this.state.pending && typeof this.state.error === 'undefined') {
      return (
        <div className="Weather" style={{ justifyContent: "flex-start" }}>
          {cityInWeatherList && (
            <div style={{ padding: "8px" }}>{this.props.data.name}</div>
          )}
          <Loader />
        </div>
      );
    } else {
      return (
        <div className="Weather">
          {this.state.error}
        </div>
      );
    }
  }
}

export default Weather;
