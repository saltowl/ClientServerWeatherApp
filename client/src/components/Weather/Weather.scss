#wind {
  margin-right: 10px;
  height: 25px;
  width: 25px;
  border: 3px white outset;
  border-radius: 50%;
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-areas:
    ". N ."
    "W windDirection E"
    ". S .";
}

#wind > div {
  align-self: center;
  justify-self: center;
  font-size: 7px;
  font-family: monospace;
}

$compass-items: N, W, windDirection, E, S;

@each $item in $compass-items {
  ##{$item} {
    grid-area: $item;
  }
}

#windDirection > div {
  font-size: 8px;
}

$rowColor: #707e89;

.Weather {
  color: #3d6787;
  width: 50%;
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  flex-direction: column;
  padding: 20px;
  box-sizing: border-box;
  font-weight: bold;
  font-size: 20px;
}

.Weather > .mainInfo,
.Weather > .mainInfo > div {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
}

.row {
  margin: 3px 0;
  padding: 7px;
  display: flex;
  flex-direction: row;
  background: $rowColor;
  color: #ffffff;
  align-items: center;
  justify-content: space-between;
  font-size: 15px;
}

.rowKey {
  font-weight: bold;
}

.rowVal {
  display: flex;
  flex-direction: row;
  align-items: center;
  font-weight: normal;
}

.temp {
  font-size: 50px;
}

.current {
  flex-direction: row-reverse !important;
  width: 100%;
  padding-right: 20px;
}

.currentName {
  align-self: flex-start;
}

.error {
  color: #fc000d;
  flex-direction: row;
  display: flex;
  justify-content: space-between;
  padding: 20px !important;
  align-items: center;
  font-weight: bold;
  width: 50%;
  box-sizing: border-box;
}

@media screen and (max-width: 715px) {
  .Weather {
    width: 100%;
    flex-direction: column !important;
  }

  .Weather > div {
    width: 100% !important;
    padding: 0 !important;
  }

  .cityMain {
    padding: 0 !important;
    order: 3;
  }

  .mainInfo:not(.current) {
    flex-wrap: wrap;
  }

  .name {
    min-width: 60%;
  }

  .temp {
    font-size: 40px !important;
    white-space: nowrap;
  }

  .error {
    width: 100%;
  }
}
