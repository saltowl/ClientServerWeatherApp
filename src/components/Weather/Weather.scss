#wind {
  margin-right: 10px;
  height: 25px;
  width: 25px;
  border: 3px white outset;
  border-radius: 50%;
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
  grid-template-columns: 1fr 1fr 1fr;
  grid-template-areas:
    ". N ."
    "W windDirection E"
    ". S .";
}

#wind > div {
  align-self: center;
  justify-self: center;
  font-size: 7px;
  font-family: monospace;
}

$compass-items: N, W, windDirection, E, S;

@each $item in $compass-items {
  ##{$item} {
    grid-area: $item;
  }
}

#windDirection > div {
    font-size: 8px;
}

$rowColor: #707e89;

.Weather {
  color: #3d6787;
  width: 50%;
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  flex-direction: column;
  padding: 20px;
  box-sizing: border-box;
  font-weight: bold;
  font-size: 20px;
}

.Weather > .mainInfo,
.Weather > .mainInfo > div {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
}

.row {
  margin: 3px 0;
  padding: 7px;
  display: flex;
  flex-direction: row;
  background: $rowColor;
  color: #ffffff;
  align-items: center;
  justify-content: space-between;
  font-size: 15px;
}

.rowKey {
  font-weight: bold;
}

.rowVal {
  display: flex;
  flex-direction: row;
  align-items: center;
  font-weight: normal;
}
