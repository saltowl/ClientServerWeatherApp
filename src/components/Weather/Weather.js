import React from 'react';
import Loader from '../Loader/Loader';
import './Weather.scss';

class Weather extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.fetchWeather();
  }
  
  render() {
    const styleWeather = this.props.data.id !== 0 ? {flexDirection: 'column'} : {flexDirection: 'row', width: '100%', fontSize: '30px'};
    const styleMainInfo = this.props.data.id !== 0 ? {flexDirection: 'row'} : {flexDirection: 'column', width: '50%', paddingRight: '20px'};
    const styleCurrentCityMain = this.props.data.id !== 0 ? {} : {flexDirection: 'row-reverse', width: '100%', paddingRight: '20px'};
    const styleCurrentCityTemp = this.props.data.id !== 0 ? {} : {fontSize: '50px'};
    const styleCurrentCityName = this.props.data.id !== 0 ? {} : {alignSelf: 'flex-start'};
    const dataRow = this.props.data.id !== 0 ? {width: '100%'} : {width: '50%', paddingLeft: '20px', boxSizing: 'border-box'};
    const styleError = {color: '#FC000D', flexDirection: 'row'};

    if (this.props.data.error) {
      const error = this.props.data.error;
      return (
        <div className="Weather" style={styleError}>
          {error.response.status === 404 &&
              <div>{this.props.data.name} isn't found</div>
          }
          <button onClick={this.props.deleteCity} className='circle'>x</button>
        </div>
      );
    } else if (this.props.data.pending === false) {
      const data = this.props.data.weather;

      const temp = `${(data.main.temp < 0 ? '' : '+')}${(data.main.temp).toFixed(1)} °C`;
      const icon = this.props.data.id !== 0 
      ? `https://openweathermap.org/img/wn/${data.weather[0].icon}.png` 
      : `https://openweathermap.org/themes/openweathermap/assets/vendor/owm/img/widgets/${data.weather[0].icon}.png`;
      const description = data.weather[0].description[0].toUpperCase() + data.weather[0].description.substr(1);
      const rot = `rotate(${data.wind.deg}deg)`;
      const windDirStyle = {
        WebkitTransform: rot,
        msTransform: rot,
        OTransform: rot,
        transform: rot
      };
      const windSpeed = `${data.wind.speed} m/s`;
      const pressure = `${(data.main.pressure * 0.75006375541921).toFixed(0)} mmHg`;
      const humidity = `${data.main.humidity}%`;
      const coords = [data.coord.lon, data.coord.lat];

      return (
        <div className="Weather" style={styleWeather}>
          <div className="mainInfo" style={styleMainInfo}>
            <div style={styleCurrentCityName}>{data.name}</div>
            <div style={styleCurrentCityMain}>
              <div style={styleCurrentCityTemp}>{temp}</div>
              <img src={icon} title={description} />
            </div>
            <div style={styleCurrentCityMain}>
              {this.props.data.id !== 0 && 
                <button onClick={this.props.deleteCity} className='circle'>x</button>
              }
            </div>
          </div>
          <div style={dataRow}>
            <div className='row'>
              <div className='rowKey'>Wind:</div>
              <div className='rowVal'>
                <div id='wind'>
                  <div id='N'>N</div>
                  <div id='W'>W</div>
                  <div id='windDirection'><div style={windDirStyle}>↑</div></div>
                  <div id='E'>E</div>
                  <div id='S'>S</div>
                </div>
                {windSpeed} 
              </div>
            </div>
            <div className='row'>
              <div className='rowKey'>Clouds:</div>
              <div className='rowVal'>{data.clouds.all}%</div>
            </div>
            <div className='row'>
              <div className='rowKey'>Pressure:</div>
              <div className='rowVal'>{pressure}</div> 
            </div>
            <div className='row'>
              <div className='rowKey'>Humidity:</div>
              <div className='rowVal'>{humidity}</div>
            </div>
            <div className='row'>
              <div className='rowKey'>Coords:</div>
              <div className='rowVal'>[{coords[0]}, {coords[1]}]</div>
            </div>
          </div>
        </div>
      );
    } else {
      return (
        <div className="Weather" style={{justifyContent: 'flex-start'}}>
            {this.props.data.id !== 0 && 
              <div style={{padding: '8px'}}>{this.props.data.name}</div>
            }
            <Loader />
        </div>
      );
    }
  }
}

export default Weather;
