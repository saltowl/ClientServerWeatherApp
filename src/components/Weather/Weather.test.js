import React from "react";
import Weather from "./Weather";
import renderer from "react-test-renderer";

const correctData = {
  base: "stations",
  clouds: { all: 75 },
  cod: 200,
  coord: { lon: 37.62, lat: 55.75 },
  dt: 1572515073,
  id: 524901,
  main: {
    temp: -2.4,
    pressure: 1021,
    humidity: 79,
    temp_min: -3,
    temp_max: -1.11
  },
  name: "Moscow",
  sys: {
    type: 1,
    id: 9027,
    country: "RU",
    sunrise: 1572496341,
    sunset: 1572530028
  },
  timezone: 10800,
  visibility: 10000,
  weather: [
    { id: 620, main: "Snow", description: "light shower snow", icon: "13d" }
  ],
  wind: { speed: 6, deg: 290 }
};

const cityNotFoundError = {
  response: {
    status: 404
  }
};

describe("Weather in the current city", () => {
  it("has been loaded", () => {
    const data = {
      id: 0,
      pending: false,
      weather: correctData
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("is pending", () => {
    const data = {
      id: 0,
      pending: true
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has not been loaded", () => {
    const data = {
      id: 0,
      pending: false,
      error: {}
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Weather in the weather list", () => {
  it("has been loaded", () => {
    const data = {
      id: 1,
      pending: false,
      weather: correctData
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("is pending", () => {
    const data = {
      id: 1,
      pending: true
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with 'city not found' error", () => {
    const data = {
      id: 1,
      pending: false,
      error: cityNotFoundError
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("has been loaded", () => {
    const data = {
      id: 1,
      pending: false,
      error: {}
    };
    const component = renderer.create(
      <Weather data={data} fetchWeather={() => {}} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
