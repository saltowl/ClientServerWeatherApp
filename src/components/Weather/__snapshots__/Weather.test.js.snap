// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather in the current city has been loaded 1`] = `
<div
  className="Weather"
  style={
    Object {
      "flexDirection": "row",
      "fontSize": "30px",
      "width": "100%",
    }
  }
>
  <div
    className="mainInfo"
    style={
      Object {
        "flexDirection": "column",
        "paddingRight": "20px",
        "width": "50%",
      }
    }
  >
    <div
      className="currentName"
    >
      Moscow
    </div>
    <div
      className="current cityMain"
    >
      <div
        className="temp"
      >
        -2.4°C
      </div>
      <img
        src="https://openweathermap.org/themes/openweathermap/assets/vendor/owm/img/widgets/13d.png"
        title="Light shower snow"
      />
    </div>
    <div
      className="current"
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "paddingLeft": "20px",
        "width": "50%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Wind:
      </div>
      <div
        className="rowVal"
      >
        <div
          id="wind"
        >
          <div
            id="N"
          >
            N
          </div>
          <div
            id="W"
          >
            W
          </div>
          <div
            id="windDirection"
          >
            <div
              style={
                Object {
                  "OTransform": "rotate(290deg)",
                  "WebkitTransform": "rotate(290deg)",
                  "msTransform": "rotate(290deg)",
                  "transform": "rotate(290deg)",
                }
              }
            >
              ↑
            </div>
          </div>
          <div
            id="E"
          >
            E
          </div>
          <div
            id="S"
          >
            S
          </div>
        </div>
        6 m/s
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Clouds:
      </div>
      <div
        className="rowVal"
      >
        75
        %
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Pressure:
      </div>
      <div
        className="rowVal"
      >
        766 mmHg
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Humidity:
      </div>
      <div
        className="rowVal"
      >
        79%
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Coords:
      </div>
      <div
        className="rowVal"
      >
        [
        37.62
        , 
        55.75
        ]
      </div>
    </div>
  </div>
</div>
`;

exports[`Weather in the current city has not been loaded 1`] = `
<div
  className="Weather error"
>
  <div>
    Poor connection. Please, try again later
  </div>
</div>
`;

exports[`Weather in the current city is pending 1`] = `
<div
  className="Weather"
  style={
    Object {
      "justifyContent": "flex-start",
    }
  }
>
  <div
    className="Loader"
  >
    <div>
      Please wait, data is loading
    </div>
    <div>
      <i
        className="fas fa-spinner fa-pulse"
      />
    </div>
  </div>
</div>
`;

exports[`Weather in the weather list has been loaded 1`] = `
<div
  className="Weather"
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <div
    className="mainInfo"
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <div
      className="name"
    >
      Moscow
    </div>
    <div
      className=" cityMain"
    >
      <div
        className=""
      >
        -2.4°C
      </div>
      <img
        src="https://openweathermap.org/img/wn/13d.png"
        title="Light shower snow"
      />
    </div>
    <div
      className=""
    >
      <button
        className="circle"
      >
        x
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Wind:
      </div>
      <div
        className="rowVal"
      >
        <div
          id="wind"
        >
          <div
            id="N"
          >
            N
          </div>
          <div
            id="W"
          >
            W
          </div>
          <div
            id="windDirection"
          >
            <div
              style={
                Object {
                  "OTransform": "rotate(290deg)",
                  "WebkitTransform": "rotate(290deg)",
                  "msTransform": "rotate(290deg)",
                  "transform": "rotate(290deg)",
                }
              }
            >
              ↑
            </div>
          </div>
          <div
            id="E"
          >
            E
          </div>
          <div
            id="S"
          >
            S
          </div>
        </div>
        6 m/s
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Clouds:
      </div>
      <div
        className="rowVal"
      >
        75
        %
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Pressure:
      </div>
      <div
        className="rowVal"
      >
        766 mmHg
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Humidity:
      </div>
      <div
        className="rowVal"
      >
        79%
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="rowKey"
      >
        Coords:
      </div>
      <div
        className="rowVal"
      >
        [
        37.62
        , 
        55.75
        ]
      </div>
    </div>
  </div>
</div>
`;

exports[`Weather in the weather list has been loaded 2`] = `
<div
  className="Weather error"
>
  <div>
    Poor connection. Please, try again later
  </div>
  <button
    className="circle"
  >
    x
  </button>
</div>
`;

exports[`Weather in the weather list is pending 1`] = `
<div
  className="Weather"
  style={
    Object {
      "justifyContent": "flex-start",
    }
  }
>
  <div
    style={
      Object {
        "padding": "8px",
      }
    }
  />
  <div
    className="Loader"
  >
    <div>
      Please wait, data is loading
    </div>
    <div>
      <i
        className="fas fa-spinner fa-pulse"
      />
    </div>
  </div>
</div>
`;

exports[`Weather in the weather list with 'city not found' error 1`] = `
<div
  className="Weather error"
>
  <div>
     isn't found
  </div>
  <button
    className="circle"
  >
    x
  </button>
</div>
`;
